akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -334,3 +339,85 @@ private[akka] final case class Expand[In, Out, Seed](seed: In ⇒ Seed, extrapol
   final override def restart(): Expand[In, Out, Seed] =
     throw new UnsupportedOperationException("Expand doesn't support restart")
 }
+
+/**
+ * INTERNAL API
+ */
+private[akka] final case class MapAsync[In, Out](parallelism: Int, f: In ⇒ Future[Out], decider: Supervision.Decider)
+  extends AsyncStage[In, Out, (Int, Try[Out])] {
+
+  type Notification = (Int, Try[Out])
+
+  private var callback: AsyncCallback[Notification] = _
+  private val elemsInFlight = FixedSizeBuffer[Try[Out]](parallelism)
+
+  override def initAsyncInput(ctx: AsyncContext[Out, Notification]): Unit = {
+    callback = ctx.getAsyncCallback()
+  }
+
+  override def onPush(elem: In, ctx: AsyncContext[Out, Notification]) = {
+    val future =
+      try f(elem)
+      catch {
+        case NonFatal(ex) if decider(ex) != Supervision.Stop ⇒ null // resume processing
+      }
+    if (future ne null) {
+      val idx = elemsInFlight.enqueue(null)
+      future.onComplete(t ⇒ callback.invoke((idx, t)))(akka.dispatch.ExecutionContexts.sameThreadExecutionContext)
+    }
+    if (elemsInFlight.isFull) ctx.holdUpstream()
+    else ctx.pull()
+  }
+
+  override def onPull(ctx: AsyncContext[Out, (Int, Try[Out])]) = {
+    @tailrec def rec(hasConsumed: Boolean): DownstreamDirective =
+      if (elemsInFlight.isEmpty && ctx.isFinishing) ctx.finish()
+      else if (elemsInFlight.isEmpty || elemsInFlight.peek == null) {
+        if (hasConsumed && ctx.isHoldingUpstream) ctx.holdDownstreamAndPull()
```

Comment:
whether or not we have previously freed up a slot in the buffer: in that case we need to push the upstream ball to get more inputs